### PR TITLE
ADR-221: Implement Sample class hierarchy and ManifestStore with JSON round-trip serialisation

### PR DESCRIPTION
**Pull request by jodavis-claude**
## Work item

ADR-221: Implement the `Sample` class hierarchy and `Manifest[S]` / `ManifestStore` in `ml/pipeline/core/` — the shared data model that every subsequent pipeline stage depends on, including JSON round-trip serialisation and the content-hash formula.

## Changes

- `ml/pipeline/__init__.py` — new empty package marker
- `ml/pipeline/core/__init__.py` — new empty package marker
- `ml/pipeline/core/sample.py` — all six dataclasses (`Sample`, `SampleWithPath`, `TextSample`, `AudioSample`, `SampleSpectrogram`, `SampleTokens`) with every field from the spec; `TextSample.id` assigned `str(uuid4())`
- `ml/pipeline/core/manifest.py` — `Manifest[S]` generic with O(1) `by_id`/`by_content_hash` index dicts; `ManifestStore.read()` (type-registry dispatch on `sample_type`) and `write()` (version-1 JSON)
- `ml/pyproject.toml` — minimal pytest config (`testpaths = ["test"]`, `pythonpath = ["."]`) enabling plain `pytest` invocation from `ml/`
- `ml/test/__init__.py` — new empty package marker
- `ml/test/pipeline/__init__.py` — new empty package marker
- `ml/test/pipeline/core/__init__.py` — new empty package marker
- `ml/test/pipeline/core/test_manifest.py` — 26 unit tests covering all four round-trip scenarios, `sample_type` dispatch, `TextSample.seed == 0` serialisation, `SampleSpectrogram.parent_id` preservation, and `AudioSample.applied_values` numeric-type fidelity

## Design decisions

- **`seed` has no dataclass default** — the spec states "PhraseVariator explicitly sets `seed=0` at construction", meaning callers pass it explicitly; a default would silently mask missing arguments in later stages.
- **Mutable dataclasses (`frozen=False`)** — the spec is silent on `frozen=`; downstream `ModifierStage` uses `Sample` instances as dict *values* (not keys), so mutability is safe and avoids unnecessary copy overhead.
- **`pyproject.toml` created in `ml/`** — resolves the known ambiguity that `validate-build` and `validate-tests` only cover .NET; enables `pytest ml/test/pipeline/core/` without manual `PYTHONPATH` manipulation.

---
_Generated by [Claude Code](https://claude.ai/code/session_01DLVaMxWuhm75L3YaMdVTFT)_

### CODE REVIEW

**Review by jodavis-claude — COMMENTED** [review]
**Priority 1 — Correctness**

- **ml/pipeline/core/manifest.py line 67:** `ManifestStore.write()` raises `IndexError` on an empty manifest. `sample_type = _SAMPLE_TYPE_KEY[type(samples[0])]` throws when `manifest.samples` is empty. `Manifest([])` is a valid object (see `test_empty_manifest`), so `write()` must either support it or raise a descriptive `ValueError("Cannot write empty manifest")` instead of an opaque index error.

- **ml/pipeline/core/manifest.py line 61:** `ManifestStore.read()` ignores the `"version"` field entirely. The JSON schema explicitly carries `"version": 1`. Without a version check, a file written by a future incompatible schema is silently deserialized as version 1 and produces corrupt objects with no diagnostic. Add `if data.get("version") != 1: raise ValueError(f"Unsupported manifest version: {data.get('version')!r}")`.

**Priority 4 — Test coverage**

- **ml/test/pipeline/core/test_manifest.py:** No test exercises `ManifestStore().write(Manifest([]), path)`. The existing `test_empty_manifest` creates `Manifest([])` and calls lookup methods but never calls `write()`, so the `IndexError` from issue 1 is not caught by the test suite. Add a test under `TestManifestLookup` (or a dedicated class) that confirms the behaviour of writing an empty manifest.

**Style (non-blocking)**

- **ml/test/pipeline/core/test_manifest.py line 4:** `import tempfile` is unused — tests use the `tmp_path` pytest fixture.
- **ml/pipeline/core/manifest.py line 61:** `_SAMPLE_TYPE_CLASS[data["sample_type"]]` raises a bare `KeyError` on unrecognised types. A `ValueError(f"Unknown sample_type: {data['sample_type']!r}")` would be more informative.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: .claude/agents/debugger.md [hunk 1]
```diff
@@ -14,6 +14,8 @@ tools:
   - Bash
   - Task
   - Skill
+  - mcp__github__issue_read
+  - mcp__github__add_issue_comment
```
Comment (jodavis) [inline .claude/agents/debugger.md]: Remove these, they're no longer necessary

File: .claude/agents/developer.md [hunk 2]
```diff
@@ -21,6 +21,7 @@ tools:
   - mcp__jira__transitionJiraIssue
   - mcp__jira__editJiraIssue
   - mcp__jira__addCommentToJiraIssue
+  - mcp__github__create_pull_request
```
Comment (jodavis) [inline .claude/agents/developer.md]: Remove, no longer necessary

File: .claude/agents/researcher.md [hunk 3]
```diff
@@ -16,6 +16,7 @@ tools:
   - mcp__microsoft-learn__microsoft_docs_search
   - mcp__microsoft-learn__microsoft_docs_fetch
   - mcp__microsoft-learn__microsoft_code_sample_search
+  - mcp__github__issue_read
```
Comment (jodavis) [inline .claude/agents/researcher.md]: Remove

File: .claude/agents/reviewer.md [hunk 4]
```diff
@@ -12,6 +12,9 @@ tools:
   - Grep
   - Bash
   - Skill
+  - mcp__github__pull_request_read
+  - mcp__github__pull_request_review_write
+  - mcp__github__update_pull_request
```
Comment (jodavis) [inline .claude/agents/reviewer.md]: Remove

File: .claude/commands/debugger-investigate.md [hunk 5]
```diff
@@ -18,12 +18,17 @@ If missing, stop and tell the caller:
 ### 1 — Fetch the GitHub issue
 
 Derive the issue number by stripping the `Issue-` prefix from the work item ID
-(e.g. `Issue-444` → `444`). Fetch the issue title, body, and all comments:
+(e.g. `Issue-444` → `444`). Fetch the issue title, body, and all comments. Prefer `gh`
```
Comment (jodavis) [inline .claude/commands/debugger-investigate.md]: The top level session will have to do this using the GitHub MCP.

File: .claude/commands/dev-team.md [hunk 6]
```diff
@@ -1,5 +1,5 @@
 ---
-description: Entry point for the dev-team agent pipeline. Routes to the correct pipeline based on the work item request, then starts the pipeline script and relays its output.
+description: Entry point for the dev-team agent pipeline. Routes to the correct pipeline based on the work item request, starts the pipeline script, and acts as the active orchestrator — handling GitHub and Jira operations on behalf of the pipeline when requested.
```
Comment (jodavis) [inline .claude/commands/dev-team.md]: The top level agent isn't the orchestrator, but it is responsible for reporting all progress. Think of it as the scrum master, getting updates from team members and updating work items, GitHub PRs and issues. 

At a high level, the orchestrator directs the team members but doesn't tell the scrum master what to do. The  scrum master just listens to progress reports and updates things appropriately.

File: .claude/commands/dev-team.md [hunk 7]
```diff
@@ -9,19 +9,21 @@ $ARGUMENTS
 
 ## Role
 
-You represent a development team. Your job is to take a work item request, choose a work 
-plan (pipeline), and then hand off the work item and the plan the pipeline script. Then you
-report progress to the user by watching the output of the script. The script is the orchestrator.
-Once you have started the script you are a passive observer.
+You are the active orchestrator for the dev-team pipeline. You start the pipeline script,
```
Comment (jodavis) [inline .claude/commands/dev-team.md]: Not the active orchestrator

File: .claude/commands/dev-team.md [hunk 8]
```diff
@@ -61,14 +63,56 @@ python -c "import sys; print(sys.platform)"
...
+   | `create_pr` | `mcp__github__create_pull_request` — use `owner`, `repo`, `title`, `head`, `base`, `body`, `draft: true` from the action; result contains the PR URL |
+   | `post_review` | `mcp__github__pull_request_review_write` — `method: "create"`, `event: "COMMENT"`, `owner`, `repo`, `pullNumber`, `body` |
+   | `get_review_threads` | `mcp__github__pull_request_read` — `method: "get_review_comments"`, `owner`, `repo`, `pullNumber`; return the full threads array |
+   | `resolve_threads` | For each `thread_id` in `thread_ids`: `mcp__github__pull_request_review_write` with `method: "resolve_thread"`, `threadId`, `owner`, `repo`, `pullNumber` |
+   | `pr_ready` | `mcp__github__update_pull_request` — `draft: false`, `owner`, `repo`, `pullNumber` |
+   | `add_reviewer` | `mcp__github__update_pull_request` — `reviewers: [reviewer]`, `owner`, `repo`, `pullNumber` |
+   | `jira_comment` | `mcp__jira__addCommentToJiraIssue` — `issue_id`, `body` from the action |
+   | `jira_transition` | `mcp__jira__transitionJiraIssue` — `issue_id`, `transition` from the action |
+
+3. **Write** the result to:
```
Comment (jodavis) [inline .claude/commands/dev-team.md]: Do not write the results to the context file. Since the team members aren't accessing the PR directly anymore, they don't need the PR URL. (It would be nice to report that to the user though). None of the other results should matter either.

File: .claude/commands/dev-team.md [hunk 9]
```diff
@@ -61,14 +63,56 @@ python -c "import sys; print(sys.platform)"
...
 - **anything else**: `tail -f <task-output-path>`
 
-Stream all output to the user as it arrives until the process exits.
+Relay each notification to the user as it arrives. When you see a line matching
+`@@ACTION_REQUIRED: <type>`, pause relaying and handle the action (see Step 4a). Resume
+monitoring afterward. Continue until the process exits.
+
+### 4a — Handle action requests
+
+When `@@ACTION_REQUIRED: <type>` appears in the pipeline output:
```
Comment (jodavis) [inline .claude/commands/dev-team.md]: Instead of looking for this token, the top level agent will just watch output from the thread and have its own rules about what to do when certain things happen. 

To get the details it needs, for things like the PR description or comment details, it can read the context file.

File: .claude/commands/prepare-pr-details.md
Comment (jodavis) [inline .claude/commands/prepare-pr-details.md:1]: Rename this from "developer-create-or" since it's not the developers skill anymore.

**Review by jodavis — CHANGES_REQUESTED** [review]
This review is focused only on the dev-team changes. Make these changes, and overall the design should be that subagents work together, with the script as the orchestrator, but they don't interact directly with GitHub or Jira. The top level agent passively watches the script's progress and makes all the updates to PRs, work items, and issues. If there's subs additional information that the top level agent needs to complete an update, we should find a way to plumb it through. But in general, the status updates front the script and the content in the context file should be enough. 

There is one other thing: since the developer and reviewer can no longer communicate through PR threads, they will have to pass a structured list of comment threads back and forth, including all the comments and their responses, who they came from, and whether or not there are resolved. That structure with have to go into the context file as well, so the top level session can see them. Might need to add an ID as well, so the top level agent can map them to GitHub PR comment threads and update the right ones. 

Also, there is a bug again that the review comments were added as a normal comment instead of a review. All the rules about PR comments that we added in the reviewer need to be transferred to the top level agent now.

**Review by jodavis-claude — COMMENTED** [review]
No prior review threads.

Two new Priority 1 (correctness/fault tolerance) issues found in ml/pipeline/core/manifest.py:

1. manifest.py:63 — data["sample_type"] is an unguarded dict subscript. The preceding version check uses data.get("version") defensively, but sample_type access raises KeyError instead of ValueError for a malformed file. Use data.get("sample_type") and raise ValueError explicitly to match the established pattern.

2. manifest.py:70-81 — write() does not validate that all samples share the same type. Python generics are erased at runtime, so Manifest([TextSample(...), AudioSample(...)]) is constructable. If passed to write(), the method silently produces a JSON file whose sample_type header matches only the first entry; on read(), _deserialise tries to construct the wrong class for the remaining entries and fails with a confusing KeyError far from the original mistake — and the corrupt file is already on disk. Add a guard at the top of write(): if any(type(s) is not type(samples[0]) for s in samples): raise ValueError(f"Manifest contains mixed sample types: {set(type(s).__name__ for s in samples)}").

All exit criteria are otherwise satisfied: all six dataclasses are present with fields matching spec lines 272-309; Manifest[S] with O(1) by_content_hash/by_id; ManifestStore read/write with discriminated dispatch; JSON schema version 1; all four round-trip tests plus dispatch, seed=0, parent_id, and numeric-type tests. The recent fix commits (version validation, ValueError for unknown sample_type, empty-manifest guard) are all correct additions.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: .claude/agents/debugger.md (hunk 1 above)
Comment (jodavis) [reply .claude/agents/debugger.md]: This change is no longer part of this PR.

File: .claude/agents/developer.md (hunk 2 above)
Comment (jodavis) [reply .claude/agents/developer.md]: This change is no longer part of this PR.

File: .claude/agents/reviewer.md (hunk 4 above)
Comment (jodavis) [reply .claude/agents/reviewer.md]: This change is no longer part of this PR.

File: .claude/agents/researcher.md (hunk 3 above)
Comment (jodavis) [reply .claude/agents/researcher.md]: This change is no longer part of this PR.

File: .claude/commands/debugger-investigate.md (hunk 5 above)
Comment (jodavis) [reply .claude/commands/debugger-investigate.md]: This change is no longer part of this PR.

File: .claude/commands/dev-team.md (hunk 6 above)
Comment (jodavis) [reply .claude/commands/dev-team.md]: This change is no longer part of this PR.

File: .claude/commands/dev-team.md (hunk 7 above)
Comment (jodavis) [reply .claude/commands/dev-team.md]: This change is no longer part of this PR.

File: .claude/commands/dev-team.md (hunk 8 above)
Comment (jodavis) [reply .claude/commands/dev-team.md]: This change is no longer part of this PR.

File: .claude/commands/dev-team.md (hunk 9 above)
Comment (jodavis) [reply .claude/commands/dev-team.md]: This change is no longer part of this PR.

File: .claude/commands/prepare-pr-details.md
Comment (jodavis) [reply .claude/commands/prepare-pr-details.md:1]: This change is no longer part of this PR.

File: ml/pipeline/core/sample.py [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+    id: stable identifier.
+      TextSample: str(uuid4()) — no file, not user-visible.
+      SampleWithPath subtypes: derived by each stage's _derive_id(); used as filename stem.
+    seed: 0 for TextSample and deterministic stages; os.urandom(8) for stochastic stages.
+    content_hash: sha256 hex digest.
+      TextSample: sha256(content.encode('utf-8')).
+      SampleWithPath subtypes: sha256(parent_content_hash + ":" + str(seed) + ":" + canonical(applied_values)).
+    """
+
+    id: str
```
Comment (jodavis) [inline ml/pipeline/core/sample.py:22]: Could we just use the hash as the ID? Ideally we shouldn't have samples with the same hash, because that means we have duplicate samples.

**Review by jodavis — CHANGES_REQUESTED** [review]

**Comment by jodavis** [comment]
@copilot Address all PR review comments

**Review by jodavis — APPROVED** [review]
